eknrenderer/eknrenderer/eknr-renderer.c

```diff
@@ -0,0 +1,1311 @@
+      (*api->error) (api, data, __LINE__, msg);
+      return 0;
+    }
+
+  if (!_renderer_section_from_ht_variant (value, api, userdata, token))
+    return 0;
+
+  return 1;
+}
+
+static void
+_renderer_set_error (mustache_api_t *api,
+                     void           *userdata,
+                     uintmax_t       lineno,
+                     const char     *msg)
+{
+  RendererMustacheData *data = userdata;
+
+  g_set_error (data->error,
+               EKNR_ERROR,
+               EKNR_ERROR_SUBSTITUTION_FAILED,
+               "Failed to perform template substitution: %s",
+               msg);
+}
+
+static mustache_api_t _renderer_mustache_data_vfuncs = {
+  .read = &_renderer_read_from_closure,
+  .write = &_renderer_write_to_closure,
+  .varget = &_renderer_var_from_ht,
+  .sectget = &_renderer_sect_from_ht,
+  .error = &_renderer_set_error
+};
+
+char *
+_renderer_render_mustache_document_internal (mustache_template_t  *tmpl,
+                                             GVariant             *variables,
+                                             GError              **error)
+{
+  g_autoptr(GVariantDict) variables_dict = g_variant_dict_new (variables);
+  g_autoptr(RendererMustacheData) data = renderer_mustache_data_new (variables_dict,
+                                                                     error,
+                                                                     NULL);
+
+  if (!mustache_render (&_renderer_mustache_data_vfuncs, (void *) data, tmpl))
+    return NULL;
+
+  return g_steal_pointer (&data->output.string);
+}
+
+/**
+ * eknr_renderer_render_mustache_document_from_file:
+ * @renderer: An #EknrRenderer
+ * @file: A #GFile specifying the location of the template file
+ * @variables: The variables and sections to use when rendering.
+ * @error: A #GError
+ *
+ * Use mustache_c to render a document, similar to
+ * eknr_renderer_render_mustache_document, but read the template
+ * from the file specified at @file. If that file has already been
+ * read, its contents will be read from the internal cache.
+ *
+ * Returns: The renderered document on success, %NULL on error.
+ */
+char *
+eknr_renderer_render_mustache_document_from_file (EknrRenderer *renderer,
+                                                  GFile        *file,
+                                                  GVariant     *variables,
+                                                  GError      **error)
+{
+  g_autofree char *uri = g_file_get_uri (file);
+  EknrRendererPrivate *priv = eknr_renderer_get_instance_private (renderer);
+  mustache_template_t *tmpl = g_hash_table_lookup (priv->cache, uri);
+  g_autofree char *contents = NULL;
+  g_autoptr(RendererMustacheData) data = NULL;
+
+  if (tmpl != NULL)
+    return _renderer_render_mustache_document_internal (tmpl, variables, error);
+
+  if (!g_file_load_contents (file, NULL, &contents, NULL, NULL, error))
+    return NULL;
+
+  data = renderer_mustache_data_new (NULL,
```

Comment:
Any reason not to pass the variables here? If you did, you could reuse `data` when rendering as well?

Reply:
Or even, is it common that we call back into this many times during the lifecycle of an EknrRenderer? Why do we not cache things in the non-file case?

Reply:
The variables aren't needed when compiling the templates, so there's no need to pass them as its an optional param. In the non-file case we don't have a key to cache by (also, this is a straight port of what the JS code did - the JS code did not cache in the non-file case).